keps/sig-network/4962-network-topology-standard/README.md

```diff
@@ -0,0 +1,954 @@
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This document proposes a standard for declaring switch network topology in Kubernetes clusters, representing the hierarchy of nodes, switches, and interconnects. In this context, a switch can refer to a physical network device or a collection of such devices with close proximity and functionality.
+
+## Motivation
+With the rise of multi-node Kubernetes workloads, especially AI workloads that demand intensive inter-node communication, scheduling pods in close network proximity is becoming essential.
+
+Some major CSPs already offer mechanisms to discover instance network topology:
+- **Amazon AWS** provides the [DescribeInstanceTopology API](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeInstanceTopology.html).
+- **Google Cloud Platform (GCP)** exposes tools via the [Google Cloud SDK](https://cloud.google.com/go/docs/reference/cloud.google.com/go/compute/latest/apiv1), allowing the retrieval of rack and cluster IDs, which can be used to reconstruct network hierarchies.
+- **Oracle Cloud Infrastructure (OCI)** offers the [Capacity Topology API](https://docs.oracle.com/en-us/iaas/tools/oci-cli/3.50.1/oci_cli_docs/cmdref/compute/capacity-topology.html), providing topology-related information for their compute nodes.
+
+Beyond CSPs, certain on-premises clusters support network topology discovery, though this capability depends on the features of the underlying switch network vendors.
+
+An open-source project, [Topograph](https://github.com/NVIDIA/topograph), has implemented these approaches and is successfully deployed in production environments.
+
+However, what remains missing is a common and standardized method to convey switch network topology information to the Kubernetes ecosystem.
+
+Currently, Kubernetes lacks a unified standard for representing network topology. This gap creates challenges for developing control plane components and applications that could leverage topology-aware features.
+
+For example, AWS has begun addressing this by introducing `topology.k8s.aws/network-node-layer-N` node labels to represent its 3-tier networking structure. However, this solution is cloud-specific and does not address broader use cases.
+
+In this KEP, we propose establishing a standardized representation of switch network topology within Kubernetes clusters.
```

Comment:
OK, but: why does this standardised way need to be part of Kubernetes itself? You need to explain that as well.

A viable alternative could be making something like NVIDIA topograph but vendor neutral, that isn't part of Kubernetes itself.

Reply:
The motivation here could also serve for a vendor-neutral alternative, not part of Kubernetes. For example, you could be proposing that https://opennetworking.org/ be the umbrella organization, not Kubernetes.

Reply:
The purpose of label standardization is to serve as a foundational step toward developing a network-aware gang scheduler for Kubernetes. By establishing a consistent way to represent cluster network topology, this initiative lays the groundwork for advanced scheduling capabilities that take network performance into account. Since this work directly enhances Kubernetes functionality, we submitted it specifically as a KEP (Kubernetes Enhancement Proposal).